README.md
# ⛔️ DEPRECATED

This package has been deprecated in favor of [@hebcal/core](https://github.com/hebcal/hebcal-es6).

# Hebcal JavaScript

Original C written by Danny Sadinoff, with portions by Michael J. Radwin.
2014 JavaScript port by Eyal Schachter (age 15).

## Description

Hebcal JS is a JavaScript API to aid working with the Hebrew calendar. It is available for both browser and Node.JS.

It can convert between Hebrew and Gregorian dates, get weekly sedra, holidays, and halachik times (_zemanim_).

You can include it in a client-side (browser) script with the `hebcal.min.js` script in the `client` subfolder (17KB gzipped). It exposes a global variable, `Hebcal`. For Node, install `hebcal` from NPM...

```sh
$ npm install hebcal
```

...and include Hebcal with `var Hebcal = require('hebcal');`

Hebcal JS is currently at version 2.3.2. It is approximately equivalent to Hebcal C 3.15.
I (Eyal) did not write Hebcal JS 1.0. 2.x is a nearly-complete rewrite of it.

The version documented here is 2.3.2.

## Releases

* 2.0.0 - 2014-02-18 (19 Adar 1 5774)
* 2.1.0 - 2014-03-16 (14 Adar 2 5774 [Purim])
* 2.1.1 - 2014-04-04 (5 Nisan 5774)
* 2.1.2 - 2014-04-04 (5 Nisan 5774)
* 2.2.0 - 2014-06-29 (1 Tamuz 5774)
* 2.2.1 - 2016-04-17 (9 Nisan 5776)
* 2.2.2 - 2016-04-18 (10 Nisan 5776)
* 2.2.3 - 2017-02-14 (18 Shvat 5777)
* 2.2.4 - 2017-05-17 (21 Iyyar 5777)
* 2.2.5 - 2017-07-02 (8 Tamuz 5777)
* 2.2.6 - 2017-10-29 (10 Cheshvan 5778)
* 2.3.0 - 2019-02-11 (6 Adar 1 5779)
* 2.3.1 - 2019-02-11 (6 Adar 1 5779)
* 2.3.2 - 2019-09-25 (25 Elul 5779)
* 2.3.3 - 2021-11-03 (28 Cheshvan 5782)


## Contributors

Hebcal JS was ported from C by Eyal Schachter (Scimonster).

The original C and JavaScript code was written by Danny Sadinoff, with contributions by Michael J. Radwin.

Various other people have made contributions on GitHub.

## License / Distribution

Copyright (C) 1994-2011 Danny Sadinoff
Portions Copyright (c) 2011 Michael J. Radwin. All Rights Reserved.

Hebcal JavaScript is distributed under the GNU Public License version 3. The original Hebcal C is under the GPLv2. The program and its source code may be freely distributed. For details, see the file COPYING in the distribution.

Remember, the GPL means you can't use this in a closed-source program! (But see [#72](https://github.com/hebcal/hebcal-js/issues/72))

If you are using this program, I'd like to hear about -- know who you are, what version you're using, and how you're using Hebcal, and anything else you'd like to tell me, so that i can adjust the program to meet users' needs.

The original Hebcal is at [hebcal/hebcal](https://github.com/hebcal/hebcal).

# API

This is where every property and method is listed.

- [CLI](https://github.com/hebcal/hebcal-js#command-line-interface)
- [`Hebcal`](https://github.com/hebcal/hebcal-js#hebcal)
- [`Hebcal.HDate`](https://github.com/hebcal/hebcal-js#new-hebcalhdateday-month-year)
- [`Hebcal.Month`](https://github.com/hebcal/hebcal-js#new-hebcalmonthmonth-year)
- [`Hebcal.GregYear`](https://github.com/hebcal/hebcal-js#new-hebcalgregyearyear-month)
- [`Hebcal.GregMonth`](https://github.com/hebcal/hebcal-js#new-hebcalgregmonthmonth-year)
- [`Hebcal.cities`](https://github.com/hebcal/hebcal-js#hebcalcities)
- [`Hebcal.holidays`](https://github.com/hebcal/hebcal-js#hebcalholidays)
- [`Hebcal.events`](https://github.com/hebcal/hebcal-js#hebcalevents)
- Utilities:
- - [`gematriya`](https://github.com/hebcal/hebcal-js#hebcalgematriyastring--num-limit)
- - [`range`](https://github.com/hebcal/hebcal-js#hebcalrangestart-end-step)

## Information on certain standards

### Descriptions

Sometimes a "description array" will be mentioned. This is an array with between 1 and 3 properties: `[Sfardit pronunciation[, Ashkenazis pronunciation[, Hebrew spelling]]]`.

Methods that fetch a value from this array expect one of the following strings: "a" for Ashkenazis, "h", for Hebrew, and anything else for Sfardit.

### Month names

Hebcal is not too particular about month names, and only checks up to the second letter. That means that, despite being totally incorrect, "cheese" will return the same value as "cheshvan". Month names work in both Hebrew and English.

Nisan, Iyyar, Sivan, Tamuz, Av, Elul, Tishrei, Cheshvan, Kislev, Tevet, Shvat, Adar (1, 2).
ניסן, אייר, סיון, תמוז, אב, אלול, תשרי, חשון, כסלו, טבת, שבט, אדר (1, 2).

### Sun times

Hebcal JS uses the [SunCalc](https://github.com/mourner/suncalc) JavaScript library to calculate sun times. They do not provide an accuracy claim, but i have found their times to be up to 3 minutes off from the accepted times. Do not rely on Zemanim until the last moment!

### Finding location in browser

Both scripts run on the exact same code, except for one thing. In the browser, Hebcal will attempt to set `Hebcal.defaultLocation` to the user's location using the [geolocation](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation) API. The `ready` event (see `Hebcal.events`) is fired once the location has been obtained, or failed to be obtained.

This will cause any webpages with the Hebcal script to request the user's permission to view their location. If they deny it, the default location will be `[0, 0]`.

To turn off this feature, use the `hebcal.noloc.min.js` script instead of the usual `hebcal.min.js` script. The unminified version is `hebcal.noloc.js`. All of these files are in the `client` folder.

### Source maps

When using Hebcal JS in the browser, it may be helpful for debugging to include a sourcemap file for browsers that support it. Both minified and unminified, regular and noloc, have a sourcemap file. This is `<js filename>.map`.

### Ordering in this documentation

Be forewarned: the ordering of properties/methods in this documentation is almost completely arbitrary. It mostly follows the order of the code, which is itself written in pretty much whatever order i decided i needed those methods. If you're looking for a specific thing, i highly recommend using your browser's find tool (ctrl+f, or cmd+f for you Mac types). If you just want to know everything about Hebcal JS, just start at the top and work your way down.

## Command line interface

Hebcal JS (as of v2.1) also provides a simple CLI. It's mainly for just testing output. If you really need an advanced CLI, use the [main Hebcal code](https://github.com/hebcal/hebcal).

You can access this CLI by installing Hebcal JS globally from NPM:

```sh
npm install -g hebcal
```

This exposes the `hebcaljs` command. This is to differentiate it from the ordinary `hebcal` command.

For usage information, run `hebcaljs -h`.

## `Hebcal`

The global variable (if running in the browser) `Hebcal` is the entrypoint to Hebcal. It is a constructor returning an object represtenting a Hebrew year, and also has many other properties.

```
new Hebcal([year[, month]]);
```

`year`, if provided, should be an `int` representing a Hebrew year, for example, `5774`. If it is falsey (includes `null`, `undefined`, `0`, `false`, etc), it defaults to the current year. If it is not falsy, and not a number, Hebcal will throw an error.

`month`, if provided, supports multiple types:

* String -- The name of a single month.
* Number -- The number of a single month (Nisan is 1, Tishrei 7).
* Array -- An array of either of the above two, or mixed. Basically, anything that can be passed to `Hebcal.Month` (see below).

If not provided, it defaults to all of the months in the given `year`.
If it is a different type, Hebcal will throw an error.

To get the full current year:

```js
var year = new Hebcal();
```

### Properties on `Hebcal` instances

#### `Hebcal.prototype.year`

Number. The year represented in the `Hebcal`. (Not actually a prototype value.)

#### `Hebcal.prototype.months`

Array of `Hebcal.Month`s. Is not intended to be used directly, but rather with the `getMonth()` and `map()` methods. (Not actually a prototype value.)

#### `Hebcal.prototype.holidays`

v2.2 and later: An object, with keys as `HDate` strings, and values are arrays of `Hebcal.holidays.Event`s. It is not recommended to use this object directly. Instead, use `HDate().holidays()` for fetching and `Hebcal().addHoliday()` for adding.

BEFORE v2.2: Array of holidays, as `Hebcal.holidays.Event`s, for the represented year.

Not actually a prototype value.

#### `Hebcal.prototype.length`

Number of days in the year. (Not actually a prototype value.)

#### `Hebcal.prototype.il`

Boolean get/set. Whether the current location is in Israel or not. This is relevant with regards to holidays.
This gets set by `Hebcal.defaultCity`, and `year.setCity()`. It's possible that you may have to set it yourself in certain circumstances. (Not actually a prototype value.)

#### `Hebcal.prototype.lat` and `long`

Number get/sets. The latitude and longitude of the current location. This is relevant with regards to getting times for dates. (Not actually prototype values.)
Default: 0.

#### `Hebcal.prototype.setCity(city)`

```js
year.setCity('Jerusalem');
```

Sets the lat/long to the given city's location. Returns the `Hebcal` it was called upon. This allows for easy method chaining.

For more information regarding cities, see `Hebcal.cities` below.

#### `Hebcal.prototype.setLocation(latitude, longitude)`

```js
year.setLocation(40.716, -74.016); // coords of New York
```

Sets the lat/long to the given location. Returns the `Hebcal` it was called upon.

#### `Hebcal.prototype.isLeapYear()`

```js
var isleap = year.isLeapYear();
```

Takes 0 arguments, returns a boolean.

#### `Hebcal.prototype.next()`

```js
var nextYear = year.next();
```

Takes 0 arguments, returns `Hebcal` representing the following year.

#### `Hebcal.prototype.prev()`

```js
var prevYear = year.prev();
```

Takes 0 arguments, returns `Hebcal` representing the preceding year.

#### `Hebcal.prototype.getMonth(month)`

```js
var tishrei = year.getMonth(7);

CHANGELOG
version 2.3.3 (2021-11-03 [28 Cheshvan 5782]))
  This package has been deprecated in favor of @hebcal/core

version 2.3.2 (2019-09-25 [25 Elul 5779])
  Hebrew language corrections [avrAhituv]

version 2.3.1 (2019-02-11 [6 Adar 1 5779])
  Fix HDate after, etc methods when setting location

version 2.3.0 (2019-02-11 [6 Adar 1 5779])
  Add isSedra method to HDate
  Fix dependency issues
  Fix bug of latitude longitude mismatch [haimkastner]
  Fix spellings -- חיי שרה [icarpis]
  Fix spellings -- כסלו
  Fix Yom Yerushalayim date
  Fix last Shabbat of year not being listed as Shabbat holiday

version 2.2.6 (2017-10-29 [10 Cheshvan 5778])
	Fix monthFromName not recognizing טבת
	Fix monthFromName treating אדר2 as Adar1
	Move source files to /src/ directory

version 2.2.5 (2017-07-02 [8 Tamuz 5777])
	Fix dependencies [shuki770]

version 2.2.4 (2017-05-17 [21 Iyyar 5777])
	Fixed previous date issue with suncalc workaround [ybudweiser]
	Add HDate.tachanun_uf() function
	Mangle variable names in minified JS

version 2.2.3 (2017-02-14 [18 Shvat 5777])
	Daf Yomi- renamed Hebrew names for Gmara according to Shas Vilna [gnudles]
	Daf Yomi- fixed errors in cycles 1-7 ... (division instead of modulo) [gnudles]
	Fix typo in Daf Yomi [asafbe]
	Don't crash on each startup [dredozubov]
	Fix year parsing from hebrew [ilyagelman]
	Write build script in JavaScript
	Tachanun is said at Mincha on 8 Tishrei -- return 7
	Don't blow up calculating molad for Tishrei

version 2.2.2 (2016-04-18 [10 Nisan 5776])
	Fix little bug caused by last-minute change breaking findParsha()

version 2.2.1 (2016-04-17 [9 Nisan 5776])
	Copy lat and long properties when using HDate().next() and prev()
	Make Hebcal().findParsha() not care about spelling format
	Fix HDate().getTishreiMonth() returning 0 in Elul
	HDate().setMonth() will wrap the year
	Tell Node not to keep the process open while listening for events

version 2.2.0 (2014-06-29 [1 Tamuz 5774])
	Add Event().routine() method
	Default HDate().holidays() to not return routine events
	Always set event based on years
	Remove holidays.IGNORE_YEAR
	Fix bug with giving a name to HDate().setMonth()
	Add HDate().candleLighting() and HDate().havdalah()
	Fix spelling of alot_hashachar zman
	Add Month().molad()
	Change names of internal methods to be shorter
	Don't include entire README in package.json (NPM will do it)
	Change Hebcal().holidays to an object instead of array
	Speed optimization for HDate().holidays()
	Remove function names (most debuggers can figure it out)
	Fix bug with candlelighting being one day off
	Don't automatically move day after sunset (was causing too many issues)

version 2.1.2 (2014-04-04 [5 Nisan 5774])
	Fix sourcemap links
	Fix GitHub link in package.json

version 2.1.1 (2014-04-04 [5 Nisan 5774])
	Change default Adar to Adar Sheini
	Add string support to new HDate()
	Fix Yom Ha'atzmaut not showing up until 1951
	Simplify cities
	Change names of Adar I and Adar II to Adar 1/2
	Fix HDate().getGregMonthObject()
	Link sourcemaps to GitHub

version 2.1 (2014-03-16 [Purim 5774])
	Add Hebcal.GregYear and Hebcal.GregMonth constructors
	Fix issue with passing a string day to new HDate(day, month[, year])
	Fix broken Adar 1/2 to new HDate(string)
	Default unspecified Adar to be 1, not 2
	Change behavior of Hebcal.prototype.map to map over individual days
	Add hallel and tachanun to HDate()s
	Add Hebcal.events property
	Setting Hebcal.defaultLocation or Hebcal.defaultCity calls an event
	Add Hebcal().getDay() method
	Fix issue with new Hebcal(-1)
	Make city names case-insensitive
	Add Ashkenazis sedra names
	Fix issue with Shabbat Mevarchim in Elul
	Add Shabbat Mevarchim and Rosh Chodesh to holidays
	Add HDate().onOrBefore()
	Add Hebcal().days() and Hebcal().filter()
	Remove some internal methods from the export
	Have HDate().getMonthObject(), getYearObject(), getGregMonthObject(), getGregYearObject() always return a value, even if unaffiliated
	Move client files to new folder
	Change cities to use lat/long decimals
	Cache results of HDate().getSedra()
	Add HDate().getParsha() alias of getSedra()
	Add Hebcal.parshiot and Hebcal.LANGUAGE
	Add Hebcal().findParsha()
	Export Hebcal.map and Hebcal.filter
	Change default havdalah offset to 42
	Fix bug with HDate(29 Adar2).next() going to the next year
	Add command line interface
	Add a few unit tests

version 2.0 (2014-02-18 [19 Adar 1 5774])
	Completely rewrote original JavaScript code


holidays.js
/*
	Hebcal - A Jewish Calendar Generator
	Copyright (C) 1994-2004  Danny Sadinoff
	Portions Copyright (c) 2002 Michael J. Radwin. All Rights Reserved.

	https://github.com/hebcal/hebcal-js

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program. If not, see <http://www.gnu.org/licenses/>.

	Danny Sadinoff can be reached at danny@sadinoff.com

	Michael Radwin has made significant contributions as a result of
	maintaining hebcal.com.

	The JavaScript code was completely rewritten in 2014 by Eyal Schachter.
 */
var c = require('./common'),
	HDate = require('./hdate'),
	gematriya = require('gematriya');

var __cache = {};

// for byte optimizations

var dayOnOrBefore = c.dayOnOrBefore,
	months = c.months,
	days = c.days,
	TISHREI = months.TISHREI,
	KISLEV = months.KISLEV,
	NISAN = months.NISAN,
	SAT = days.SAT,
	getDay = 'getDay',
	abs = 'abs',
	Shabbat = 'Shabbat',
	Shabbos = 'Shabbos';

function Chanukah(day) {
	return ['Chanukah: Candle ' + day, 0, 'חנוכה: נר ' + gematriya(day)];
}

function CHM(desc) {
	return [desc[0] + ' (CH"M)', desc[1] ? desc[1] + ' (CH"M)' : desc[1], desc[2] ? desc[2] + ' )חה"ם(' : desc[2]];
}

function Sukkot(day) {
	return ['Sukkot: ' + day, 'Succos: ' + day, 'סוכות יום ' + gematriya(day)];
}

function Pesach(day) {
	return ['Pesach: ' + day, 0, 'פסח יום ' + gematriya(day)];
}

var USER_EVENT          = 1,
	LIGHT_CANDLES       = 2,
	YOM_TOV_ENDS        = 4,
	CHUL_ONLY           = 8, // chutz l'aretz (Diaspora)
	IL_ONLY             = 16, // b'aretz (Israel)
	LIGHT_CANDLES_TZEIS = 32;

exports.masks = {
	USER_EVENT         : USER_EVENT,
	LIGHT_CANDLES      : LIGHT_CANDLES,
	YOM_TOV_ENDS       : YOM_TOV_ENDS,
	CHUL_ONLY          : CHUL_ONLY,
	IL_ONLY            : IL_ONLY,
	LIGHT_CANDLES_TZEIS: LIGHT_CANDLES_TZEIS
};

function Event(date, desc, mask) {
	var me = this;
	me.date = new HDate(date);
	me.desc = typeof desc != 'object' ? [desc] : desc;

	me.USER_EVENT          = !!( mask & USER_EVENT           );
	me.LIGHT_CANDLES       = !!( mask & LIGHT_CANDLES        );
	me.YOM_TOV_ENDS        = !!( mask & YOM_TOV_ENDS         );
	me.CHUL_ONLY           = !!( mask & CHUL_ONLY            );
	me.IL_ONLY             = !!( mask & IL_ONLY              );
	me.LIGHT_CANDLES_TZEIS = !!( mask & LIGHT_CANDLES_TZEIS  );
}

Event.prototype.is = function(date, il) {
	date = new HDate(date), myDate = this.date;
	if (arguments.length < 2) {
		//il = Event.isIL;
		il = date.il;
	}
	if (date.getDate() != myDate.getDate() || date.getMonth() != myDate.getMonth()) {
		return false;
	}
	if (date.getFullYear() != myDate.getFullYear()) {
		return false;
	}
	if (il && this.CHUL_ONLY || !il && this.IL_ONLY) {
		return false;
	}
	return true;
};

Event.prototype.masks = function() {
	var me = this;
	return (me.USER_EVENT          && USER_EVENT)    |
		   (me.LIGHT_CANDLES       && LIGHT_CANDLES) |
		   (me.YOM_TOV_ENDS        && YOM_TOV_ENDS)  |
		   (me.CHUL_ONLY           && CHUL_ONLY)     |
		   (me.IL_ONLY             && IL_ONLY)       |
		   (me.LIGHT_CANDLES_TZEIS && LIGHT_CANDLES_TZEIS);
};

Event.prototype.getDesc = function(o) {
	return c.LANG(this.desc, o);
};

Event.prototype.candleLighting = function() {
	var date = this.date;
	if (this.LIGHT_CANDLES) {
		return new Date(date.sunset() - (Event.candleLighting * 60 * 1000));
	} else if (this.LIGHT_CANDLES_TZEIS) {
		return date.getZemanim().tzeit;
	}
	return null;
};

Event.prototype.havdalah = function() {
	if (this.YOM_TOV_ENDS) {
		return new Date(this.date.sunset().getTime() + (Event.havdalah * 60 * 1000));
	}
	return null;
};

Event.prototype.routine = (function(){
	function routine(){
		return !!~routine.names.indexOf(this.getDesc('s'));
	}
	routine.names = [Shabbat, 'Erev ' + Shabbat];
	return routine;
})();

Event.isIL = false;

Event.candleLighting = 18;

Event.havdalah = 42;

exports.Event = Event;

exports.year = function(year) {
	if (__cache[year]) {
		return __cache[year];
	}

	var	RH = new HDate(1, TISHREI, year),
		pesach = new HDate(15, NISAN, year),
		tmpDate;

	var h = {};

	function add(ev) {
		if (Array.isArray(ev)) {
			ev.forEach(function(e){
				add(e);
			});
		} else {
			if (h[ev.date]) {
				h[ev.date].push(ev);
			} else {
				h[ev.date] = [ev];
			}
		}
	}

	Object.defineProperty(h, 'add', {value: add});

	add([ // standard holidays that don't shift based on year
		new Event(
			RH,
			['Rosh Hashana 1', 0, 'ראש השנה א\''],
			LIGHT_CANDLES_TZEIS
		), new Event(
			new HDate(2, TISHREI, year),
			['Rosh Hashana 2', 0, 'ראש השנה ב\''],
			YOM_TOV_ENDS
		), new Event(
			new HDate(3 + (RH[getDay]() == days.THU), TISHREI, year), // push off to SUN if RH is THU
			['Tzom Gedaliah', 0, 'צום גדליה'],
			0
		), new Event(
			new HDate(9, TISHREI, year),
			['Erev Yom Kippur', 0, 'ערב יום כיפור'],
			LIGHT_CANDLES
		), new Event( // first SAT after RH
			new HDate(dayOnOrBefore(SAT, 7 + RH[abs]())),
			[Shabbat + ' Shuva', Shabbos + ' Shuvah', 'שבת שובה'],
			0
		), new Event(
			new HDate(10, TISHREI, year),
			['Yom Kippur', 0, 'יום כיפור'],
			YOM_TOV_ENDS
		), new Event(
			new HDate(14, TISHREI, year),
			['Erev Sukkot', 'Erev Succos', 'ערב סוכות'],
			LIGHT_CANDLES
		), new Event(
			new HDate(15, TISHREI, year),
			Sukkot(1),
			LIGHT_CANDLES_TZEIS | CHUL_ONLY
		), new Event(
			new HDate(15, TISHREI, year),
			Sukkot(1),
			YOM_TOV_ENDS | IL_ONLY
		), new Event(
			new HDate(16, TISHREI, year),
			Sukkot(2),
			YOM_TOV_ENDS | CHUL_ONLY
		), new Event(
			new HDate(16, TISHREI, year),
			CHM(Sukkot(2)),
			IL_ONLY
		), new Event(
			new HDate(17, TISHREI, year),
			CHM(Sukkot(3)),
			0
		), new Event(
			new HDate(18, TISHREI, year),
			CHM(Sukkot(4)),
			0
		), new Event(
			new HDate(19, TISHREI, year),
			CHM(Sukkot(5)),
			0
		), new Event(
			new HDate(20, TISHREI, year),
			CHM(Sukkot(6)),
			0
		), new Event(
			new HDate(21, TISHREI, year),
			['Sukkot: 7 (Hoshana Raba)', 'Succos: 7 (Hoshana Raba)', 'סוכות יום ז\' )הושנע רבה('],
			LIGHT_CANDLES
		), new Event(
			new HDate(22, TISHREI, year),
			['Shmini Atzeret', 'Shmini Atzeres', 'שמיני עצרת'],
			LIGHT_CANDLES_TZEIS | CHUL_ONLY
		), new Event(
			new HDate(22, TISHREI, year),
			['Shmini Atzeret / Simchat Torah', 'Shmini Atzeres / Simchas Torah', 'שמיני עצרת / שמחת תורה'],
			YOM_TOV_ENDS | IL_ONLY
		), new Event(
			new HDate(23, TISHREI, year),
			['Simchat Torah', 'Simchas Torah', 'שמחת תורה'],
			YOM_TOV_ENDS | CHUL_ONLY
		), new Event(
			new HDate(24, KISLEV, year),
			['Erev Chanukah', 0, 'ערב חנוכה'],
			0
		), new Event(
			new HDate(25, KISLEV, year),
			Chanukah(1),
			0
		), new Event(
			new HDate(26, KISLEV, year),
			Chanukah(2),
			0
		), new Event(
			new HDate(27, KISLEV, year),
			Chanukah(3),
			0
		), new Event(
			new HDate(28, KISLEV, year),
			Chanukah(4),
			0
		), new Event(
			new HDate(29, KISLEV, year),
			Chanukah(5),
			0
		), new Event(
			new HDate(30, KISLEV, year), // yes, i know these are wrong
			Chanukah(6),
			0
		), new Event(
			new HDate(31, KISLEV, year), // HDate() corrects the month automatically
			Chanukah(7),
			0
		), new Event(
			new HDate(32, KISLEV, year),
			Chanukah(8),
			0
		), new Event(
			new HDate(15, months.SHVAT, year),
			["Tu B'Shvat", 0, 'ט"ו בשבט'],
			0
		), new Event(
			new HDate(dayOnOrBefore(SAT, pesach[abs]() - 43)),
			[Shabbat + ' Shekalim', Shabbos + ' Shekalim', 'שבת שקלים'],
			0
		), new Event(
			new HDate(dayOnOrBefore(SAT, pesach[abs]() - 30)),
			[Shabbat + ' Zachor', Shabbos + ' Zachor', 'שבת זכור'],
			0
		), new Event(
			new HDate(pesach[abs]() - (pesach[getDay]() == days.TUE ? 33 : 31)),
			["Ta'anit Esther", "Ta'anis Esther", 'תענית אסתר'],
			0
		), new Event(
			new HDate(13, months.ADAR_II, year),
			['Erev Purim', 0, 'ערב פורים'],
			0
		), new Event(
			new HDate(14, months.ADAR_II, year),
			['Purim', 0, 'פורים'],
			0
		), new Event(
			new HDate(15, months.ADAR_II, year),
			['Shushan Purim', 0, 'שושן פורים'],
			0
		), new Event(
			new HDate(dayOnOrBefore(SAT, pesach[abs]() - 14) - 7),
			[Shabbat + ' Parah', Shabbos + ' Parah', 'שבת פרה'],
			0
		), new Event(
			new HDate(dayOnOrBefore(SAT, pesach[abs]() - 14)),
			[Shabbat + ' Hachodesh', Shabbos + ' Hachodesh', 'שבת החודש'],
			0
		), new Event(
			new HDate(dayOnOrBefore(SAT, pesach[abs]() - 1)),
			[Shabbat + ' HaGadol', Shabbos + ' HaGadol', 'שבת הגדול'],
			0
		), new Event(
			// if the fast falls on Shabbat, move to Thursday
			pesach.prev()[getDay]() == SAT ? pesach.onOrBefore(days.THU) : new HDate(14, NISAN, year),
			["Ta'anit Bechorot", "Ta'anis Bechoros", 'תענית בכורות'],
			0
		), new Event(
			new HDate(14, NISAN, year),
			['Erev Pesach', 0, 'ערב פסח'],
			LIGHT_CANDLES
		), new Event(
			new HDate(15, NISAN, year),
			Pesach(1),
			LIGHT_CANDLES_TZEIS | CHUL_ONLY
		), new Event(
			new HDate(15, NISAN, year),
			Pesach(1),
			YOM_TOV_ENDS | IL_ONLY
		), new Event(
			new HDate(16, NISAN, year),
			Pesach(2),
			YOM_TOV_ENDS | CHUL_ONLY
		), new Event(
			new HDate(16, NISAN, year),
			CHM(Pesach(2)),
			IL_ONLY
		), new Event(
			new HDate(16, NISAN, year),
			['Start counting Omer', 0, 'התחלת ספירת העומר'],
			0
		), new Event(
			new HDate(17, NISAN, year),
			CHM(Pesach(3)),
			0
		), new Event(
			new HDate(18, NISAN, year),
			CHM(Pesach(4)),
			0
		), new Event(
			new HDate(19, NISAN, year),
			CHM(Pesach(5)),
			0
		), new Event(
			new HDate(20, NISAN, year),
			CHM(Pesach(6)),
			LIGHT_CANDLES
		), new Event(
			new HDate(21, NISAN, year),
			Pesach(7),
			LIGHT_CANDLES_TZEIS | CHUL_ONLY
		), new Event(
			new HDate(21, NISAN, year),
			Pesach(7),
			YOM_TOV_ENDS | IL_ONLY
		), new Event(
			new HDate(22, NISAN, year),
			Pesach(8),
			YOM_TOV_ENDS | CHUL_ONLY
		), new Event(
			new HDate(14, months.IYYAR, year),
			['Pesach Sheni', 0, 'פסח שני'],
			0
		), new Event(
			new HDate(18, months.IYYAR, year),
			["Lag B'Omer", 0, 'ל"ג בעומר'],
			0
		), new Event(
			new HDate(5, months.SIVAN, year),
			['Erev Shavuot', 'Erev Shavuos', 'ערב שבועות'],
			LIGHT_CANDLES
		), new Event(
			new HDate(6, months.SIVAN, year),
			['Shavuot 1', 'Shavuos 1', 'שבועות א\''],
			LIGHT_CANDLES_TZEIS | CHUL_ONLY
		), new Event(
			new HDate(6, months.SIVAN, year),
			['Shavuot', 'Shavuos', 'שבועות'],
			YOM_TOV_ENDS | IL_ONLY
		), new Event(
			new HDate(7, months.SIVAN, year),
			['Shavuot 2', 'Shavuos 2', 'שבועות ב\''],
			YOM_TOV_ENDS | CHUL_ONLY
		), new Event(
			new HDate(dayOnOrBefore(SAT, new HDate(1, TISHREI, year + 1)[abs]() - 4)),
			['Leil Selichot', 'Leil Selichos', 'ליל סליחות'],
			0
		), new Event(
			new HDate(29, months.ELUL, year),
			['Erev Rosh Hashana', 0, 'ערב ראש השנה'],
			LIGHT_CANDLES
		)
	]);

	tmpDate = new HDate(10, months.TEVET, year);
	if (tmpDate[getDay]() == SAT) {
		tmpDate = tmpDate.next();
	}
	add(new Event(
		tmpDate,
		["Asara B'Tevet", 0, 'עשרה בטבת'],
		0
	));

	if (c.LEAP(year)) {
		add(new Event(
			new HDate(14, months.ADAR_I, year),
			['Purim Katan', 0, 'פורים קטן'],
			0
		));

		add(new Event(
			new HDate(15, months.ADAR_I, year),
			['Shushan Purim Katan', 0, 'שושן פורים קטן'],
			0
		));
	}

	if (year >= 5711) { // Yom HaShoah first observed in 1951
		tmpDate = new HDate(27, NISAN, year);
		/* When the actual date of Yom Hashoah falls on a Friday, the
		 * state of Israel observes Yom Hashoah on the preceding
		 * Thursday. When it falls on a Sunday, Yom Hashoah is observed
		 * on the following Monday.
		 * http://www.ushmm.org/remembrance/dor/calendar/
		 */

		if (tmpDate[getDay]() == days.FRI) {
			tmpDate = tmpDate.prev();
		} else if (tmpDate[getDay]() == days.SUN) {
			tmpDate = tmpDate.next();
		}

		add(new Event(
			tmpDate,
			['Yom HaShoah', 0, 'יום השואה'],
			0
		));
	}

	add(atzmaut(year));

	if (year >= 5727) { // Yom Yerushalayim only celebrated after 1967
		add(new Event(
			new HDate(29, months.IYYAR, year),
			['Yom Yerushalayim', 0, 'יום ירושלים'],
			0
		));
	}

	tmpDate = new HDate(17, months.TAMUZ, year);
	if (tmpDate[getDay]() == SAT) {
		tmpDate = tmpDate.next();
	}
	add(new Event(
		tmpDate,
		["Shiva-Asar B'Tamuz", 0, "צום יז' בתמוז"],
		0
	));

	tmpDate = new HDate(9, months.AV, year);
	if (tmpDate[getDay]() == SAT) {
		tmpDate = tmpDate.next();
	}

	add(new Event(
		new HDate(dayOnOrBefore(SAT, tmpDate[abs]())),
		[Shabbat + ' Chazon', Shabbos + ' Chazon', 'שבת חזון'],
		0
	));

	add(new Event(
		tmpDate.prev(),
		["Erev Tish'a B'Av", 0, 'ערב תשעה באב'],
		0
	));

	add(new Event(
		tmpDate,
		["Tish'a B'Av", 0, 'תשעה באב'],
		0
	));

	add(new Event(
		new HDate(dayOnOrBefore(SAT, tmpDate[abs]() + 7)),
		[Shabbat + ' Nachamu', Shabbos + ' Nachamu', 'שבת נחמו'],
		0
	));

	for (var day = 6; day < c.daysInYear(year); day += 7) {
		add(new Event(
			new HDate(dayOnOrBefore(SAT, new HDate(1, TISHREI, year)[abs]() + day)),
			[Shabbat, Shabbos, 'שבת'],
			YOM_TOV_ENDS
		));

		add(new Event(
			new HDate(dayOnOrBefore(days.FRI, new HDate(1, TISHREI, year)[abs]() + day)),
			['Erev ' + Shabbat, 'Erev ' + Shabbos, 'ערב שבת'],
			LIGHT_CANDLES
		));
	}

	for (var month = 1; month <= c.MONTH_CNT(year); month++) {
		if ((month == NISAN ? c.daysInMonth(c.MONTH_CNT(year - 1), year - 1) :
				c.daysInMonth(month - 1, year)) == 30) {
			add(new Event(
				new HDate(1, month, year),
				['Rosh Chodesh 2', 0, "ראש חודש ב'"],
				0
			));

			add(new Event(
				new HDate(30, month - 1, year),
				['Rosh Chodesh 1', 0, "ראש חודש א'"],
				0
			));
		} else if (month !== TISHREI) {
			add(new Event(
				new HDate(1, month, year),
				['Rosh Chodesh', 0, 'ראש חודש'],
				0
			));
		}

		if (month == months.ELUL) {
			continue;
		}

		add(new Event(
			new HDate(29, month, year).onOrBefore(SAT),
			[Shabbat + ' Mevarchim', Shabbos + ' Mevorchim', 'שבת מברכים'],
			0
		));
	}

	return __cache[year] = h;
};

function atzmaut(year) {
	if (year >= 5708) { // Yom HaAtzma'ut only celebrated after 1948
		var tmpDate = new HDate(1, months.IYYAR, year), pesach = new HDate(15, NISAN, year);

		if (pesach[getDay]() == days.SUN) {
			tmpDate.setDate(2);
		} else if (pesach[getDay]() == SAT) {
			tmpDate.setDate(3);
		} else if (year < 5764) {
			tmpDate.setDate(4);
		} else if (pesach[getDay]() == days.TUE) {
			tmpDate.setDate(5);
		} else {
			tmpDate.setDate(4);
		}

		return [new Event(
			tmpDate,
			['Yom HaZikaron', 0, 'יום הזיכרון'],
			0
		), new Event(
			tmpDate.next(),
			["Yom HaAtzma'ut", 0, 'יום העצמאות'],
			0
		)];
	}
	return [];
}
exports.atzmaut = atzmaut;
